blockchain/validator_set.go

```diff
@@ -0,0 +1,273 @@
+package blockchain
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+
+	acm "github.com/hyperledger/burrow/account"
+	rpc_core "github.com/tendermint/tendermint/rpc/core"
+	ctypes "github.com/tendermint/tendermint/rpc/core/types"
+	tm_types "github.com/tendermint/tendermint/types"
+)
+
+type ValidatorSet interface {
+	TotalPower() int
+	SetMaximumPower(maximumPower int)
+	MaximumPower() int
+	AdjustPower(height int64) error
+	Validators() []acm.Validator
+	SetLeavers() []acm.Validator
+	JoinToTheSet(validator acm.Validator) error
+	LeaveFromTheSet(validator acm.Validator) error
+	IsValidatorInSet(address acm.Address) bool
+	JSONBytes() ([]byte, error)
+}
+
+type validatorListProxy interface {
+	Validators(height int64) (*ctypes.ResultValidators, error)
```

Comment:
Is this working around the fact we don't have a clean way to get the current validator set from ABCI? This is dubious and probably induces a race conditions. We really need to work against a set of validators and a particular commit height and have all of this happen while we hold the commit lock.

Reply:
For two reason I defined the proxy interface here:
1- writing test independently
2- In future we might change the way we get validator list. It might be get directly from ABCI.

Reply:
Yep I just wanted to check that was the case - interface is a good idea but the current implementation through the RPC is problematic and will lead to a race condition - we should definitely be getting this through the ABCI in the future.

Reply:
I also used rpc_core for checking sortition transaction. That part also should change. Thanks